### PR TITLE
Locsys condtion for beams

### PR DESCRIPTION
**Pull request by knarfnitram**
Adds the `LOCSYS` condition to meshpy. 

I have created only a test case for a `POINT LOCSYS` with beams. 
@isteinbrecher  Do we need additional tests for: 

- line 
- surface 
- volume ?

### CODE REVIEW

File: meshpy/boundary_condition.py [hunk 1]
```diff
@@ -80,6 +80,7 @@ def from_dat(cls, bc_key, line, **kwargs):
 
         if bc_key in (
             mpy.bc.dirichlet,
+            mpy.bc.locsys,
```
Comment (isteinbrecher) [inline meshpy/boundary_condition.py:83]: Ordering should match the one in `meshpy/conf.py`. I would put it at the end.

File: tests/testing_meshpy.py [hunk 2]
```diff
@@ -1943,6 +1943,55 @@ def test_meshpy_display_pyvista(self):
...
+                bc_type=mpy.bc.dirichlet,
+            )
+        )
+
+        # Add local sys condition.
+        input_file.add(
+            BoundaryCondition(
+                beam_set["end"],
+                "ROTANGLE 0.0 0.0 0.1 FUNCT 0 0 0 USEUPDATEDNODEPOS 0 USECONSISTENTNODENORMAL 0",
+                bc_type=mpy.bc.locsys,
```
Comment (isteinbrecher) [inline tests/testing_meshpy.py:1988]: Since we have quite an elaborate large rotational implementation within meshpy, I suggest a different approach here:
Lets create a class `LocSysCondition` (preferably in the `four_c` subdirectory) that takes a `Rotation` as input argument.

Should look something like this:
```python
class LocSysCondition(BoundaryCondition):

    def __init__(
        self,
        geometry,
        rotation,
        *,
        **kwargs,
    ):
        super().__init__(geometry,
             "ROTANGLE {} {} {} FUNCT 0 0 0 USEUPDATEDNODEPOS 0 USECONSISTENTNODENORMAL 0".format(*rotation.get_rotation_vector()),
             bc_type=mpy.bc.locsys, **kwargs)
```

File: meshpy/boundary_condition.py (hunk 1 above)
Comment (knarfnitram) [reply meshpy/boundary_condition.py]: Thanks, for catching that.

File: tests/testing_meshpy.py [hunk 3]
```diff
@@ -1943,6 +1944,49 @@ def test_meshpy_display_pyvista(self):
         mesh = self.create_beam_to_solid_conditions_model()
         _plotter = mesh.display_pyvista(is_testing=True, resolution=3)
 
+    def test_meshpy_locsys_condition(self):
```
Comment (isteinbrecher) [inline tests/testing_meshpy.py:1947]: Docstrings should start directly after the three quotations:
```
        """Test case for point locsys condition for beams.
        The testcase is similar to beam3r_herm2line3_static_locsys.dat, but with simpler material.
        """
```

File: tests/testing_meshpy.py [hunk 4]
```diff
@@ -1943,6 +1944,49 @@ def test_meshpy_display_pyvista(self):
...
+            )
+        )
+
+        input_file.add(
+            BoundaryCondition(
+                beam_set["end"],
+                "NUMDOF 9 ONOFF 1 0 0 0 0 0 0 0 0 VAL 1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 FUNCT 1 0 0 0 0 0 0 0 0",
+                bc_type=mpy.bc.dirichlet,
+            )
+        )
```
Comment (isteinbrecher) [inline tests/testing_meshpy.py:1982]: Why do we need these here?

Comment (knarfnitram) [reply tests/testing_meshpy.py:1982]: I recreated the test `beam3r_herm2line3_static_locsys` just with a simpler material, therefore I added the additional bc on the same points (to see if this is a problem, but it isn't).
One can delete it, but then I would also remove the reference at the top to the `beam3r_herm2line3_static_locsys`.

Comment (isteinbrecher) [reply tests/testing_meshpy.py:1982]: Then lets leave this and add a comment stating that the other BC are there to check that the combination works

**Comment by isteinbrecher** [comment]
@knarfnitramwhat is the status of this merge request? Is something still missing?

**Comment by knarfnitram** [comment]
I have updated the branch, so I think it's ready to be merged.

**Review by isteinbrecher — APPROVED** [review]

File: meshpy/four_c/locsys_condition.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+"""
+
+
+# MeshPy modules
+from ..conf import mpy
+from ..boundary_condition import BoundaryCondition
+
+
+class LocSysCondition(BoundaryCondition):
+    """
```
Comment (isteinbrecher) [inline meshpy/four_c/locsys_condition.py:42]: „This object represents a locsys condition in 4C. It allows to rotate the local coordinate system used to apply Dirichlet boundary conditions“

File: tests/testing_four_c.py [hunk 6]
```diff
@@ -175,6 +179,48 @@ def create_model():
...
+            )
+        )
+        # Add additional dirichlet boundary condtion to check if combination with locsys condition works.
+        input_file.add(
+            BoundaryCondition(
+                beam_set["end"],
+                "NUMDOF 9 ONOFF 1 0 0 0 0 0 0 0 0 VAL 1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 FUNCT 1 0 0 0 0 0 0 0 0",
+                bc_type=mpy.bc.dirichlet,
+            )
+        )
```
[the remark this answers is not in the document]
Comment (isteinbrecher) [reply tests/testing_four_c.py:299]: Ignore this commen.

File: meshpy/four_c/locsys_condition.py (hunk 5 above)
Comment (isteinbrecher) [reply meshpy/four_c/locsys_condition.py:42]: Line formatting seems off here. I think once this is fixed we can merge.

Comment (knarfnitram) [reply meshpy/four_c/locsys_condition.py:42]: Thanks, I thought that black would have fixed this.

Comment (isteinbrecher) [reply meshpy/four_c/locsys_condition.py:42]: `black` does not change doc strings.

**Review by isteinbrecher — APPROVED** [review]

**Comment by isteinbrecher** [comment]
@knarfnitram you should also be able to merge in this repository now.
